### PR TITLE
Fix/564: Async error when reading from JWKS endpoint

### PR DESCRIPTION
**Pull request by dvanmali**
# Description

Fixes to the JWKS Config file using async and promise behavior.

# Issue

- Converted to async behavior to await a response from the JWKS Endpoint
- Errors from both the verification function and JWKS endpoint now propagate back to the user instead of getting stuck in the callback.
- Added TCK test for Config file as requested from @oskarhane.
- Fixed failing integration test 505.int.test.tx

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [ ] (n/a) New files have copyright header
 - [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by darrellwarde** [comment]
Hey @dvanmali, I'm having some confusion between this and #564 regarding the configuration for this.

In #564, it was [documented](https://github.com/neo4j/graphql/pull/564/files?authenticity_token=DcCavgPetCrFDVEn0ZkOCFrupoGYkwyQSHBmW41280fbWDs1DTvXyPMW46B3EuXy9LIJ%2Fk64S6%2Fz7ApJjUBRqQ%3D%3D&file-filters%5B%5D=.adoc&file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-3544bee8679eb9d1099e7b770df35769b421dfe412100476a42768f5dc5bb475) as `jwksEndpoint` but then [the interface](https://github.com/neo4j/graphql/pull/564/files?authenticity_token=DcCavgPetCrFDVEn0ZkOCFrupoGYkwyQSHBmW41280fbWDs1DTvXyPMW46B3EuXy9LIJ%2Fk64S6%2Fz7ApJjUBRqQ%3D%3D&file-filters%5B%5D=.adoc&file-filters%5B%5D=.json&file-filters%5B%5D=.ts#diff-17e24e319550c2497da6fbf92f48861f6b9edeb061331375d0831ba414fc98a9) defined it as `jwkEndpoint`.

This PR seems to move to the common ground of `jwksEndpoint` - was this a typo in the previous branch? We obviously now enter slightly dubious territory that this is now _technically_ a breaking change, but we might just have to draw a blind eye to this in the case of a typo.

Additionally, this additional configuration needs adding to the [API reference](https://neo4j.com/docs/graphql-manual/current/api-reference/neo4jgraphql/#api-reference-neo4jgraphql-input-neo4jgraphqlconfig-neo4jgraphqljwt) in [this file](https://neo4j.com/docs/graphql-manual/current/api-reference/neo4jgraphql/#api-reference-neo4jgraphql-input-neo4jgraphqlconfig-neo4jgraphqljwt).

**Comment by darrellwarde** [comment]
Additionally, as part of this PR, could you please correct the typos introduced in the previous PR on [this line](https://github.com/neo4j/graphql/pull/564/files#diff-3544bee8679eb9d1099e7b770df35769b421dfe412100476a42768f5dc5bb475R8)?

This should read:

```
- `jwksEndpoint` (recommended) - The OpenID Public Key Endpoint that verifies the JWT. The https://auth0.com/docs/security/tokens/json-web-tokens/json-web-key-sets[JSON Web Key Sets] used under the OpenID Configuration are typically stored at an endpoint similar to \https://YOUR_DOMAIN/.well-known/jwks.json
```

Just note the "recommended" and "are" instead of "is".

**Comment by dvanmali** [comment]
This PR fixes all `jwkEndpoint` to `jwksEndpoint` which formally stands for Json Web Key Sets. I noticed that typo and fixed `s` and its integration test in object-path.int.test.ts. The prior PR created an async error during every call, so the feature would result in an error/break every time anyways.

Just added a commit to add those recommended documentations fixes.

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-578.surge.sh

File: docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc:105]: If you take a look at the preview site commented by the bot, you'll see that you've broken the table a bit here. You _should_ be able to apply the suggestion below to fix this. I've also removed "recommended" as I'm of the opinion that an API reference should be purely factual and not opinionated.

```suggestion
|`jwksEndpoint` +
 +
 Type: `string`
|The OpenID Public Key Set Endpoint used to verify JWT tokens. The JWKS used to store the verification information is typically stored at an endpoint similar to \https://YOUR_DOMAIN/.well-known/jwks.json.

|`secret` +
 +
 Type: `string`
|The secret used to encode JWT tokens.

```

File: packages/graphql/src/auth/get-jwt.ts
Comment (darrellwarde) [inline packages/graphql/src/auth/get-jwt.ts:111]: Could this be moved outside of the `getJWT` function as ESLint is suggesting? Even just in the same file is fine, it just makes it a bit clearer.

**Review by darrellwarde — CHANGES_REQUESTED** [review]
Looking good, just a couple of comments!

File: packages/graphql/src/auth/get-jwt.ts
Comment (dvanmali) [reply packages/graphql/src/auth/get-jwt.ts]: Sure that's fine, moved it outside the function but not allowing it as a file export

**Comment by dvanmali** [comment]
Fixed all recommendations @darrellwarde, please double check all docs before merging :) thanks!

File: docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc
Comment (darrellwarde) [reply docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc]: Just fixing some formatting here and removing "Recommended" as requested - I've just committed this one straight in due to it being super simple.

File: docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/api-reference/neo4jgraphql.adoc:96]:
```suggestion
|The OpenID Public Key Set Endpoint used to verify JWT tokens. The JWKS used to store the verification information is typically stored at an endpoint similar to \https://YOUR_DOMAIN/.well-known/jwks.json.
```

**Comment by darrellwarde** [comment]
> * Fixed failing integration test 505.int.test.ts

Trying to figure out what was failing here originally but struggling due to the number of changes in the file. Any chance you could give an overview? Were the failures caused by changes introduced in this branch?

Broadly speaking, happy enough with the work in this branch and the general support for JWKS endpoints, but I do take issue with the fact that it appears we have no tests that actually validate the functionality. It may be too complex, but is there any way you know of standing up a local JWKS server for testing purposes and testing against it?

**Comment by dvanmali** [comment]
When running tests on my local machine 505.int.test.ts was the only one failing, while running the test suite locally so I researched the issue. I'm not sure exactly when this failure arose, it was always a problem with me testing locally. 

TLDR: It seemed to be a problem with how we sent over GraphQL syntax and not an issue with this library.

Previously, the following [query](https://github.com/neo4j/graphql/blob/8a856be4f36266efc58c84c50ffffa0834e777f7/packages/graphql/tests/integration/issues/505.int.test.ts#L163) was run:
```javascript
const mutation = `
            query PeopleWorkspaceAndPages($userId: ID!, $workspaceId: ID!) {
                people(where: { id: $userId } ) {
                    createdPages {
                        id
                    }
                }
                workspaces(where: { id: $workspaceId })
                {
                    pages {
                        id
                    }
                }
                pages(where: { workspace: { id: $workspaceId } } ) {
                    id
                }
                allPages: pages {
                    id
                }
            }
        `;
```

However, this type of syntax is not a valid GraphQL query. GraphQL has 1 starting node and queries from there. The above tries to query from 3 starting nodes which doesn't make sense: people, workspaces, pages. Instead, we need to separate the query as follows: 

```javascript
        const usersQuery = `
            query Users($userId: ID!) {
                users(where: { id: $userId }) {
                    id,
                    authId,
                    createdPages {
                        id
                    }
                }
            }
        `;
        const workspacesQuery = `
            query Workspaces($workspaceId: ID!) {
                workspaces(where: { id: $workspaceId }) {
                    id,
                    pages {
                        id
                    }
                }
            }
        `;
        const pagesQuery = `
            query Pages($workspaceId: ID!) {
                pages(where: { workspace: { id: $workspaceId } }) {
                    id
                }
            }
        `;
        const allPagesQuery = `
            query allPages {
                pages {
                    id
                }
            }
        `;
```

<details open>
<summary>Another valid query if you want to group everything together, only difference is that you'd to specify the function you would use</summary>
<br>
```javascript
        const allQueries = `
            query Users($userId: ID!) {
                users(where: { id: $userId }) {
                    id,
                    authId,
                    createdPages {
                        id
                    }
                }
            }
            query Workspaces($workspaceId: ID!) {
                workspaces(where: { id: $workspaceId }) {
                    id,
                    pages {
                        id
                    }
                }
            }
            query Pages($workspaceId: ID!) {
                pages(where: { workspace: { id: $workspaceId } }) {
                    id
                }
            }
            query allPages {
                pages {
                    id
                }
            }
        `;
`;
```
</details>

Now, the queries are separated into functions with one starting node and only the parameter used. The test still performs the exact test addressed by #505, but with proper syntax.

**Comment by dvanmali** [comment]
> but I do take issue with the fact that it appears we have no tests that actually validate the functionality. It may be too complex, but is there any way you know of standing up a local JWKS server for testing purposes and testing against it?

Took me some digging but I found a mock-jwks library that can do this up, I'll get working on this then notify when finished

**Comment by dvanmali** [comment]
Alright @darrellwarde completed the integration test with mock-jwt. See commit `af5efb1`

**Comment by darrellwarde** [comment]
> Alright @darrellwarde completed the integration test with mock-jwt. See commit `af5efb1`

Brilliant, thanks @dvanmali, I'll aim to get this reviewed today! 🙂

File: packages/graphql/tests/integration/directives/auth/jwks-endpoint.int.test.ts
Comment (darrellwarde) [inline packages/graphql/tests/integration/directives/auth/jwks-endpoint.int.test.ts:99]: These particular tests don't appear to be testing anything of this project at all?

**Review by darrellwarde — COMMENTED** [review]
This looks good, thanks for adding those tests, they will give us a lot more confidence in this feature.

Just a question about the first few tests that have been added?

File: packages/graphql/tests/integration/directives/auth/jwks-endpoint.int.test.ts
Comment (dvanmali) [reply packages/graphql/tests/integration/directives/auth/jwks-endpoint.int.test.ts:99]: You are correct, these test the settings of the jwksMock server itself. I kept it to ensure the server is up and running as intended before the actual tests. Mostly because I didn't how a local mock server will react in your GitHub CI/CD on-prem tests. But if you want to clean up clutter, we can remove it... your call

**Review by darrellwarde — APPROVED** [review]
Brilliant, this is looking good to go to me! Many thanks for your patience during review. 🙂

**Pull request merged**
